pkg/hydrate/controller.go

```diff
@@ -99,20 +104,38 @@ func (h *Hydrator) Run(ctx context.Context) {
 					klog.Errorf("failed to complete the rendering execution for commit %q: %v",
 						srcCommit, err)
 				}
-			} else if doneCommit(h.DonePath.OSPath()) != srcCommit {
-				// If the commit has been processed before, regardless of success or failure,
-				// skip the hydration to avoid repeated execution.
-				// The rehydrate ticker will retry on the failed commit.
-				hydrateErr = h.hydrate(srcCommit, syncPath)
-				if err := h.complete(srcCommit, hydrateErr); err != nil {
-					klog.Errorf("failed to complete the rendering execution for commit %q: %v", srcCommit, err)
+			} else {
+				doneCommit := extractCommit(h.DonePath.OSPath())
+				readyToRenderFile := h.ReconcilerSignalDir.Join(cmpath.RelativeSlash(ReadyToRenderFile)).OSPath()
+				readyToRenderCommit := extractCommit(readyToRenderFile)
+				switch {
+				case doneCommit == srcCommit:
+					// no-op
+					// If the commit has been processed before, skip the hydration to avoid repeated execution regardless of success or failure.
+					// The rehydrate ticker will retry on the failed commit.
+				case doneCommit != srcCommit && readyToRenderCommit != srcCommit:
+					klog.Warningf("skip hydration as readyToRenderCommit does not match srcCommit, want %s, got %s", srcCommit, readyToRenderCommit)
+				case doneCommit != srcCommit && readyToRenderCommit == srcCommit:
+					// Only proceed to rendering if the ready-to-render file contains the srcCommit
+					hydrateErr = h.hydrate(srcCommit, syncPath)
+					if err := h.complete(srcCommit, hydrateErr); err != nil {
+						klog.Errorf("failed to complete the rendering execution for commit %q: %v", srcCommit, err)
+					}
 				}
 			}
 			runTimer.Reset(h.PollingPeriod) // Schedule re-run attempt
 		}
 	}
 }
 
+// isReadyToRender returns true if commit in ready-to-render file is the same as
+// commit acquired from source
+func (h *Hydrator) isReadyToRender(srcCommit string) (string, bool) {
```

Comment:
Is this function still needed?

Reply:
Yes, I split it out so that we can do unit test on it.

Reply:
It is only used in the unit test. If it is not used by the controller, we don't need to unit test this function.

Reply:
Ah I forgot the logic in Run has changed. Just updated with cleanup.